### PR TITLE
Rotem commit for schedule task

### PR DESCRIPTION
**Pull request by rotem443**

### CODE REVIEW

File: app/models/product.rb [hunk 1]
```diff
@@ -3,4 +3,12 @@ class Product < ActiveRecord::Base
 
   has_and_belongs_to_many :battles
   has_many :votes
+
+  def retriveVotes
+  	if (self.fake_votes)
+  		return self.fake_votes
+  	else
+  		return self.votes
+	end
```
Comment (odedharth) [inline app/models/product.rb]: you can write this like this as well:
`self.fake_votes || self.votes`

File: app/models/product.rb [hunk 2]
```diff
@@ -3,4 +3,12 @@ class Product < ActiveRecord::Base
 
   has_and_belongs_to_many :battles
   has_many :votes
+
+  def retriveVotes
```
Comment (odedharth) [inline app/models/product.rb]: should be - retrive_votes

File: app/models/battle.rb [hunk 3]
```diff
@@ -37,4 +39,19 @@ def set_category
 
     BattlesCategories.create(category_id: cat.id, battle_id: self.id)
   end
+
+  def scheduleCreate(first_product_id, second_product_id) 
+    if self.save
```
Comment (odedharth) [inline app/models/battle.rb]: create_schedule

File: app/controllers/api/v1/products_controller.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+				render json: @product.id, status: :created
+			else
+				render json: @product.errors, status: :unprocessable_entity
+			end
+		end
+
+		def destroy_many
+			@products = Product.where(:id => params[:delete_list])
+			
+			if @products.destroy_all
```
Comment (odedharth) [inline app/controllers/api/v1/products_controller.rb]: why do we need this?

File: app/controllers/api/v1/battles_controller.rb [hunk 5]
```diff
@@ -13,10 +13,32 @@ def index
...
+
+      first_product_id = params[:battle][:products_attributes][0][:product_id]
+      second_product_id = params[:battle][:products_attributes][1][:product_id]
+      schedule_time = params[:battle][:schedule_time]
+
+      result = @battle.scheduleCreate(first_product_id,
+                                      second_product_id)
+        
+
+      if (schedule_time)
```
Comment (odedharth) [inline app/controllers/api/v1/battles_controller.rb]: no need for the brackets

File: Gemfile [hunk 6]
```diff
@@ -1,5 +1,5 @@
 source 'https://rubygems.org'
-ruby '2.2.3'
+ruby '2.1.5'
```
Comment (odedharth) [inline Gemfile]: why you lowered it down?
I think we should even try to up to ruby 2.3

Comment (odedharth) [inline Gemfile]: why you lowered it down?
I think we should even try to up to ruby 2.3

File: app/controllers/api/v1/battles_controller.rb [hunk 7]
```diff
@@ -13,10 +13,32 @@ def index
...
+        Rufus::Scheduler.singleton.at schedule_time do
+          @battle.update_attribute(:published, true)
+
```
Comment (odedharth) [inline app/controllers/api/v1/battles_controller.rb]: no need for the space here

File: app/controllers/api/v1/battles_controller.rb [hunk 8]
```diff
@@ -13,10 +13,32 @@ def index
...
-        render nothing: true, status: :created
+      @battle.update_attribute(:published, (!params[:battle][:schedule_time]))
+
+      first_product_id = params[:battle][:products_attributes][0][:product_id]
+      second_product_id = params[:battle][:products_attributes][1][:product_id]
+      schedule_time = params[:battle][:schedule_time]
+
+      result = @battle.scheduleCreate(first_product_id,
+                                      second_product_id)
+
```
Comment (odedharth) [inline app/controllers/api/v1/battles_controller.rb]: two spaces instead of one

File: app/controllers/api/v1/feed_controller.rb [hunk 9]
```diff
@@ -1,8 +1,13 @@
 module Api
   class Api::V1::FeedController < ActionController::API
+    require 'date'
 
     def index
-      battles = Battle.feed(params[:user_id], params[:category_name])
+      future_time = DateTime.parse(params[:future_time])
+
+      #format('YYYY-MM-DD HH:mm:ss')
```
Comment (odedharth) [inline app/controllers/api/v1/feed_controller.rb]: remove comments and spaces

File: app/controllers/api/v1/products_controller.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
+module Api	
+class Api::V1::ProductsController < ActionController::API
+		def index
+	  		@products = Product.all
+
+	  		render json: @products
```
Comment (odedharth) [inline app/controllers/api/v1/products_controller.rb]: lots of spaces in this page that could be removed

File: app/models/product.rb (hunk 1 above)
Comment (rotem443) [inline app/models/product.rb]: Fixed this

File: app/models/product.rb (hunk 2 above)
Comment (rotem443) [inline app/models/product.rb]: Fixed

File: app/models/battle.rb (hunk 3 above)
Comment (rotem443) [inline app/models/battle.rb]: Fixed

File: app/controllers/api/v1/products_controller.rb (hunk 4 above)
Comment (rotem443) [inline app/controllers/api/v1/products_controller.rb]: For the angular website

File: app/controllers/api/v1/search_controller.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
+module Api
+  class Api::V1::SearchController < ActionController::API
+  	require 'amazon/ecs'
+
```
Comment (odedharth) [inline app/controllers/api/v1/search_controller.rb]: its good that you created a search controller
but all of the code related to the integration with amazon should be moved into a ruby class called Search put this class in the lib directory (auto load this class http://stackoverflow.com/questions/19098663/auto-loading-lib-files-in-rails-4)
the class should have a method named 'search_from amazon' that get a query and returns the results

the goal is that in the controller you simply need to do :
`@productList = Search.search_from_amazon(params[:query])`

File: app/controllers/api/v1/search_controller.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  	def index
+		
+		res = Amazon::Ecs.item_search(params[:query], { :response_group => 'Large', :country => 'us', :search_index => 'All'})
+
+		@productList = res.items.map do |item|
+			nPrice = -1
+			item_attributes = item.get_element('ItemAttributes')	 		
+
+			# Product that you can instantly buy
```
Comment (odedharth) [inline app/controllers/api/v1/search_controller.rb]: can do this instead in one line:

`nPrice = item.get_element('ListPrice').get('FormattedPrice')  if item.get_element('ListPrice') && item.get_element('ListPrice').get('FormattedPrice')`

Comment (odedharth) [inline app/controllers/api/v1/search_controller.rb]: there is also a better way to do this with try (http://everydayrails.com/2011/04/28/rails-try-method.html) like this:
`
nPrice = item.get_element('ListPrice').get('FormattedPrice') if item.get_element('ListPrice').try(:get,'FormattedPrice')`

File: app/controllers/api/v1/battles_controller.rb (hunk 5 above)
Comment (rotem443) [inline app/controllers/api/v1/battles_controller.rb]: Changed to code

File: app/models/battle.rb [hunk 13]
```diff
@@ -21,14 +23,17 @@ def self.feed(user_id, category_name)
           battles = self.all
         end
 
-        return battles.where("battles.id NOT IN (SELECT DISTINCT(battle_id) FROM votes WHERE user_id = #{user_id})")
+        return battles.where("battles.id NOT IN (SELECT DISTINCT(battle_id) FROM votes WHERE user_id = #{user_id})").where(published: true)
```
Comment (odedharth) [inline app/models/battle.rb]: return battles.where("battles.id NOT IN (SELECT DISTINCT(battle_id) FROM votes WHERE user_id = #{user_id}) AND published = true")

File: Gemfile (hunk 6 above)
Comment (rotem443) [inline Gemfile]: Yes, you're right. 
my local version is 2.1.5 and i was to lazy to install the new one.
i will install the 2.3 version now

File: config/initializers/schedule.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+	@battles = Battle.includes(:products).last(3)
+	@battles.each do |battle|
+		#Rails.logger.debug "winner_precent #{battle.winner_precent}"
+
+		if (battle.winner_precent)
+
+		  	first_fake_vote  = battle.products[0].fake_votes + (rand(battle.winner_precent..100) / 10)
+		  	second_fake_vote = battle.products[1].fake_votes + (rand(0..(100-battle.winner_precent)) / 10)
+
+		  	 # Rails.logger.debug "first_fake_vote #{first_fake_vote}"
```
Comment (odedharth) [inline config/initializers/schedule.rb]: remove comments

File: app/controllers/api/v1/battles_controller.rb (hunk 7 above)
Comment (rotem443) [inline app/controllers/api/v1/battles_controller.rb]: Fixed

File: app/controllers/api/v1/battles_controller.rb (hunk 8 above)
Comment (rotem443) [inline app/controllers/api/v1/battles_controller.rb]: Fixed

File: config/initializers/schedule.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+		  	 # Rails.logger.debug "second_fake_vote #{second_fake_vote}"
+
+		  	 # Rails.logger.debug "first_fake_vote_real #{battle.products[0].fake_votes}"
+		  	 # Rails.logger.debug "second_fake_vote #{battle.products[1].fake_votes}"
+
+			battle.products[0].update_attribute(:fake_votes, first_fake_vote)
```
Comment (odedharth) [inline config/initializers/schedule.rb]: you can do it with one sql query with update_attributes method

File: app/controllers/api/v1/feed_controller.rb (hunk 9 above)
Comment (rotem443) [inline app/controllers/api/v1/feed_controller.rb]: Fixed

File: app/controllers/api/v1/search_controller.rb (hunk 11 above)
Comment (rotem443) [inline app/controllers/api/v1/search_controller.rb]: after gal help, fixed

File: app/controllers/api/v1/search_controller.rb (hunk 12 above)
Comment (rotem443) [inline app/controllers/api/v1/search_controller.rb]: Fixed, changed to the second version

**Pull request merged**